### ISSUE TITLE
[objects.layout] Rename layout to grid

### ISSUE THREAD

**Issue by nicoqh** [inuitcss/inuitcss#152]
I think the layout object is more properly described as a grid.

I usually create my own `layout` object which includes a header, a navigation, a mobile drawer and the main content holder. These encapsulate the layout of the page and are related to each other in terms of functionality and theming. I believe `o-layout` would be a good class name for this.

The `grid` object, however, I use to divide (sections of) the page into smaller parts. In my projects these are unrelated to the layout object described above.

My rationale for renaming the object is as follows:

* You only use one grid system, and "grid" poses fewer risks for conflicts
* The definition of "grid" fits perfectly for this object
* A grid > cell makes more sense than a layout -> item
* You would probably refer to the current layout object as a "grid" if you were to explain it to someone (see previous issues and source comments in objects.layout and utilities.widths).

Would love some feedback on this :)

**Comment by tymondesigns** [comment]
There are some valid points in here.

If it were to be renamed, then personally I would suggest keeping the `__item {}` naming convention as it keeps everything succinct and consistent.

**Comment by inlikealion** [comment]
I think the name grid is intentionally avoided here because of the way it works. It's not a 12 column grid with cells, it's a way of dividing available space in certain ratios. Different mental modal, and I support the distinction in naming.

**Comment by dnixx** [comment]
I think the different mental model is due implementation details, not how the grid behaves. It's not Bootstrap's 12 column grid, but it does result in an X-column grid with cells and gutters.

If not, I think it would imply that a "grid system" could be imported alongside the layout system, which would make the distinction useful, but I don't see a reason to do that. If you want another _implementation_, say, flexbox, you'd end up with a grid class anyway.

Nonetheless I think "layout" is too generic for an object that either creates a grid or is a sectioning tool for horizontal space.

**Comment by herzinger** [comment]
It IS a 12 column grid if you want it to be. It's a grid system no matter how you look at it.

**Comment by herzinger** [comment]
I have a mixin that generates grid systems. You can pass the name of the object as a parameter, and you can generate simple (like a 12 columns one) or recursive (like the 1-5 grids), so it also solves a discussion I saw somewhere else I can't find, about which kind of grid to have by default. With some tweaking it all could be fully configurable, in a project level.

If you guys are interested I could make a PR.

**Comment by csshugs** [comment]
> I think the name grid is intentionally avoided here because of the way it works. It's not a 12 column grid with cells, it's a way of dividing available space in certain ratios. 

Exactly. 
It's not a grid. It's a way to **lay out** elements in fractional dividers on the page. It's in fact a different mental modal.

> Nonetheless I think "layout" is too generic for an object that either creates a grid or is a sectioning tool for horizontal space.

The aim of an object is to be as generic as possible, also considering the naming of the class names. When you chose a CSS framework, you always have to watch out for the classes the framework brings with it to prevent naming collisions. You either have to change the framework or adapt the naming of your own classes if you want to prevent naming collisions. 

There are other classes in inuitcss that are even more generic like:

* `.box`
* `.block`
* `.media`
* `.wrapper`

That's what makes an object, IMO. Being generic in source code and in naming.

**Comment by herzinger** [comment]
@csshugs so what you're saying is something along the lines of: it's not a grid system per se, but an object that you can make a grid system out of.

**Comment by inlikealion** [comment]
@herzinger It's _not_ a grid system, because one doesn't have to think in every available horizontal space “Ok, I have 12 columns I need to account for here”. At any given space you can choose how to subdivide it individually. It's a very different mental modal to 960 Grid System and the like. In Foundation you choose how many columns and then have to account for that many columns in every “row“ of horizontal space.

**Comment by herzinger** [comment]
Yes, but you CAN use it like a traditional grid system, by outputing only 12 columns or so.  It even has advantages in a design standpoint.

**Comment by rowild** [comment]
@herzinger So you want it to be renamed to grid, too?

**Comment by herzinger** [comment]
@rowild no, I was satisfied with @csshugs' explanation. But I have something I'm planing to make a plugin of that has something to do with that, so this definition is useful.

**Comment by nicoqh** [comment]
I'm aware of the differences between grid systems like Bootstrap's and inuitcss' layout system. Both systems divide horizontal space into chunks by using predefined percentage-based classes, whether they're named u-1/2 or col-sm-6[/12], and they're both accompanied by gutters. The total doesn't have to amount to 100% in either implementation.

I personally don't think this represents a completely different mental model, nor a difference in outcome, which is one of the reasons I proposed the name change. If the "layout" system had a completely different job and couldn't replace the "grid" system, I wouldn't have made the proposal.

@csshugs, I absolutely agree that objects should be generic. My concern is that the UIs I make have very different layouts (stacked, holy grail etc.) which I represent with a "layout" class (how the page is laid out), but all of them also need to divide units of horizontal space into chunks, which I represent with a "grid" class.

In other words, the layout of the page will be different for each app, but not the idea of dividing horizontal space into predefined ratios (and I would refer to this as a grid regardless of implementation detail because the ratios and the invisible dividing "lines" are predefined, just like Bootstrap).

This distinction makes more sense to me, but because inuitcss is very modular it's not big deal to import another grid/layout system if I find that need :)

**Comment by herzinger** [comment]
>  If the "layout" system had a completely different job and couldn't replace the "grid" system, I wouldn't have made the proposal.

The layout object can replace a grid system, but a traditional grid system can't replace the layout object, is what it all comes to in my opinion. It's more versatile.

@nicoqh that aside, I'm working in a plugin/standalone tool that might solve your problem. I think during the weekend I'll have a repo up.

**Comment by csshugs** [comment]
@nicoqh 

> I usually create my own layout object which includes a header, a navigation, a mobile drawer and the main content holder. These encapsulate the layout of the page and are related to each other in terms of functionality and theming. I believe o-layout would be a good class name for this.

Actually, what you described here would rather be a component than an object—it's a concrete piece of your UI, hence `.c-layout` would be more appropriate. Nonetheless, having a component class with the name `.c-layout` and an object with the class name `.o-layout` could be a bit confusing, giving that these two classes have absolutely nothing to do with each other.

So, one suggestion would be to rename your class to something like `.c-page-layout` which distinguishes better from `.o-layout`. Or you don't use the inuitcss object.layout module and write yourself an object with your favored class `.o-grid`. I would suggest the former solution, because that's what you do elsewhere, anyway. As [I stated above](https://github.com/inuitcss/inuitcss/issues/152#issuecomment-249178119), when you are using a 3rd-party framework like inuitcss, there will always be classes that you need to look out for and cannot use, because they are reserved by the framework itself. The way one normally goes is: Find yourself a CSS framework you like, then write CSS yourself to add the components you need on top of the framework and use classes for that, which aren't existing yet in the framework. 
The way you describe your problem here is like, I'm going to build some media gallery for example, name the class to `.media` and then asking the framework for renaming the media-object to `.some-other-class`, because it conflicts with my custom class.

I really understand your problem, but one of the few things even inuitcss has to be opinionated about is class names. We decided to call it `.o-layout` for the reasons pointed out in the comments above. We are sorry if the class names we've chosen conflict with your existing naming best-practices. And it's still Ok, of course to bring things like this up when you are concerned about naming or think, the naming is off. But in this case we will purposely keep the naming as it is.

**Comment by nicoqh** [comment]
> [...] name the class to .media and then asking the framework for renaming the media-object to .some-other-class, because it conflicts with my custom class.

Couldn't agree more, but understand that the crux of my argument was the naming of the object regardless of local conflicts (I don't use `o-layout`).

Thanks for taking the time to discuss this, and for the good suggestions :)
